### ISSUE TITLE
[question] how MsMarco v5 models are trained?

### ISSUE THREAD

**Issue by pommedeterresautee** [UKPLab/sentence-transformers#1167]
Hi,

I have seen the new V5 models pushed to the Huggingface Hub and it appears that bi-encoder scores are very high (close to SOTA).
In particular those trained with the new margin MSE loss.
I am wondering if you can share how they have been trained?

Can you confirm you are using the following script: https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder_margin-mse.py

Moreover, regarding the number of epochs, I understand from the hub that it has been trained during 30 epochs.

Can you share when the progress appeared, in particular, did you get most of the value during the first n epochs, and choose an arbitrary large number of epochs to be sure to not miss some MRR points behind, or is it more linear (each epoch bringing a little progress compared to the precedent epoch)?
Can you share what kind of hardware have been used, and training times?

Last question, about the dot product thing, I did some test on a large private dataset (> 500K pairs of cleaned q/a in legal domain), most of the passages have around 380 tokens, and scores with dot product + MultipleNegativesRankingLoss were slightly lower than with cosine + MultipleNegativesRankingLoss. I was wondering if this change on the MsMarco was important in terms of scores.

Thank you for all information you can provide.

**Comment by nreimers** [comment]
Hi,
yes, that is the clean version of the training script.

The new models contain a train_script.py:
https://huggingface.co/sentence-transformers/msmarco-distilbert-dot-v5/blob/main/train_script.py

Which is the actual training script.

1) It was trained for 30 epochs using hard negatives from other system.
2) Then hard negatives were mined for that particular system
3) Then another 30 epochs were trained on the hard negatives from step 2.


Step 3 is not really necessary.


Here the performance numbers for the msmarco-BERT-model:

Step 1)
Number of train steps | MRR@10
110k | 35,5
180k | 36,12
260k | 36,54
410k | 36,98
471k | 37,19  (30 epochs)


Step 3 has the following performance numbers:
130k | 37,67
200k | 37,75
330k | 37,8
430k | 37,92
471k | 38,08 (30 epochs)



I think training any further does not make so much sense. We are already overfitting on MSMARCO, i.e. any value above ~36 MRR@10 is mainly due to overfitting on MSMARCO. If you test these models on other retrieval tasks or on TREC-DL datasets, you don't large differences there.

**Comment by pommedeterresautee** [comment]
thank you @nreimers it's very clear. Can you share the hardware used for training and the time it took approximatively?

**Comment by nreimers** [comment]
Hardware was a single V100 GPU. DistilBERT models took less than 1 day per step (so less than 3 days in total). BERT models maybe two times longer, so less than 6 days in total.

**Comment by pommedeterresautee** [comment]
thanks a lot for these precisions

**Comment by nirmal2k** [comment]
> 1. It was trained for 30 epochs using hard negatives from other system.
> 2. Then hard negatives were mined for that particular system
> 3. Then another 30 epochs were trained on the hard negatives from step 2.

@nreimers. How are hard negatives mined for that particular system in step2?

**Comment by nreimers** [comment]
You encode all passages. 
Then you encode all training queries.
For each training queries, you find the top 50 closest passages in vector space. 
Each mined (query, passage) pair is then classified by the cross-encoder

**Comment by nirmal2k** [comment]
Thanks for clearing that.
In the [train_script](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/ms_marco/train_bi-encoder_margin-mse.py), I see that in the class MSMARCODataset, postive samples and negatives samples are popped once used. So that does mean the actual positives (qrels.train) are being used only once in all 30 epochs?

**Comment by nreimers** [comment]
They are popped and then appended at the end again. So we will iterate over them

**Comment by nirmal2k** [comment]
Oh I'm sorry, I didn't notice the append.

**Comment by nirmal2k** [comment]
Oh I'm sorry, I didn't notice the append.
